library/src/main/scala/scala/scalajs/js/FinalizationRegistry.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.scalajs.js
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation.JSGlobal
+
+/** <span class="badge badge-ecma2021" style="float: right;">ECMAScript 2021</span>
+ *
+ *  A FinalizationRegistry object lets you request a callback when an object is garbage-collected.
+ *
+ *  FinalizationRegistry provides a way to request that a cleanup callback get called at some point
+ *  when an object registered with the registry has been reclaimed (garbage-collected).
+ *  (Cleanup callbacks are sometimes called finalizers.)
+ *
+ *  MDN
+ */
+@js.native
+@JSGlobal
+class FinalizationRegistry[T](finalizer: js.Function1[T, Unit]) {
+  /** The register() method registers an object with a FinalizationRegistry instance so that if
+   *  the object is garbage-collected, the registry's callback may get called.
+   *
+   *  @param theObject The target object to register.
+   *  @param heldValue The value to pass to the finalizer for this object. This cannot be the target object.
+   *  @param unregistrationToken A token that may be used with the `unregister`` method later to
+   *                             unregister the target object. If provided (and not `undefined`),
+   *                             this must be an object. If not provided, the target cannot be unregistered.
+   */
+  def register(theObject: Any, heldValue: T, unregistrationToken: Any): Unit = js.native
+
+  /** The register() method registers an object with a FinalizationRegistry instance so that if
+   *  the object is garbage-collected, the registry's callback may get called.
+   *
+   *  @param theObject The target object to register.
+   *  @param heldValue The value to pass to the finalizer for this object. This cannot be the target object.
+   */
+  def register(theObject: Any, heldValue: T): Unit = js.native
+
+  /** The unregister unregisters a target object from a FinalizationRegistry instance.
+   *
+   *  When a target object has been reclaimed, it is no longer registered in the registry.
+   *  There is no need to all unregister in your cleanup callback. Only call unregister if
+   *  you haven't received a cleanup callback and no longer need to receive one.
+   *
+   *  @param unregistrationToken The token used with the register method when registering the target object.
+   */
+  def unregister(unregistrationToken: Any): Unit = js.native
```

Comment:
I wonder if we should make `FinalizationRegistry` parametric in the type of the `unregistrationToken`s as well :thinking:

Reply:
We might also have a type parameter for the type of objects that are registered, i.e., the type of `theObject`. I'm not sure whether going that far actually helps, though.

Regardless of what we choose to model with type parameters, they can all be contravariant, unless I'm mistaken.